core/unit_test/TestGraph.hpp

```diff
@@ -420,112 +420,243 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), empty_graph) {
-      },
-      [&](AllocateDataEvent alloc) {
-        if (alloc.name != alloc_label)
-          return MatchDiagnostic{
-              false, {"Allocation name mismatch (got " + alloc.name + ')'}};
-        if (alloc.size < functor_t::count)
-          return MatchDiagnostic{
-              false,
-              {"Allocation size mismatch (expected at least " +
-               std::to_string(functor_t::count) + " but got " +
-               std::to_string(alloc.size) + ')'}};
-        ptr      = alloc.ptr;
-        ptr_size = alloc.size;
-        return MatchDiagnostic{true};
-      }));
+template <typename T>
+struct TestGraphSizedFunctorLaunch : public ::testing::Test {
+  static constexpr auto launch_mechanism = T::value;
 
-  EXPECT_TRUE(static_cast<bool>(graph));
-  graph->instantiate();  // NOLINT(bugprone-unchecked-optional-access)
+  static constexpr size_t threshold =
+      TestGraphSizedFunctorLaunchTraits<launch_mechanism>::threshold;
 
-  // Fencing the default execution space instance, as the node policy
-  // was created without giving an instance (it used the default one).
-  TEST_EXECSPACE{}.fence(
-      "Ensure that kernel dispatch to global memory is finished "
-      "before submission.");
+  static constexpr bool allocates_driver_memory_buffer =
+      TestGraphSizedFunctorLaunchTraits<
+          launch_mechanism>::allocates_driver_memory_buffer;
 
-  graph->submit(ex);  // NOLINT(bugprone-unchecked-optional-access)
-  ASSERT_TRUE(contains(ex, data, functor_t::count));
-
-  ASSERT_TRUE(validate_event_set(
-      [&]() { graph.reset(); },
-      [&](DeallocateDataEvent dealloc) {
-        if (dealloc.name == alloc_label && dealloc.ptr == ptr &&
-            dealloc.size == ptr_size)
-          return MatchDiagnostic{true};
+  // Ensure that the launch mechanism chosen for a given functor size works.
+  template <typename ExecSpace>
+  void test_sized_functor_launch(const ExecSpace& exec) {
+    using view_t =
+        Kokkos::View<int, ExecSpace, Kokkos::MemoryTraits<Kokkos::Atomic>>;
+
+    using functor_t = SizedFunctor<view_t, threshold + 1>;
+
+#if defined(KOKKOS_ENABLE_CUDA)
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::Cuda>) {
+      using driver_t = Kokkos::Impl::ParallelFor<
+          functor_t,
+          Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::Impl::IsGraphKernelTag>,
+          Kokkos::Cuda>;
+      static_assert(
+          Kokkos::Impl::DeduceCudaLaunchMechanism<driver_t>::launch_mechanism ==
+          launch_mechanism);
+    }
+#endif
+
+#if defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::HIP>) {
+      using driver_t = Kokkos::Impl::ParallelFor<
+          functor_t,
+          Kokkos::RangePolicy<Kokkos::HIP, Kokkos::Impl::IsGraphKernelTag>,
+          Kokkos::HIP>;
+      static_assert(
+          Kokkos::Impl::DeduceHIPLaunchMechanism<driver_t>::launch_mechanism ==
+          launch_mechanism);
+    }
+#endif
```

Comment:
I don;t think we should do that.
This is 20 lines of code I don;t want to read and that leak implementation details.